src/librustc_mir/transform/elaborate_drops.rs

```diff
@@ -257,18 +268,38 @@ impl<'a, 'b, 'tcx> DropElaborator<'a, 'tcx> for Elaborator<'a, 'b, 'tcx> {
         })
     }
 
-    fn array_subpath(&self, path: Self::Path, index: u32, size: u32) -> Option<Self::Path> {
-        dataflow::move_path_children_matching(self.ctxt.move_data(), path, |p| {
-            match p {
-                &Projection {
-                    elem: ProjectionElem::ConstantIndex{offset, min_length: _, from_end: false}, ..
-                } => offset == index,
-                &Projection {
-                    elem: ProjectionElem::ConstantIndex{offset, min_length: _, from_end: true}, ..
-                } => size - offset == index,
-                _ => false
-            }
-        })
+    fn array_subpaths(&self, path: Self::Path, size: u64)
+                      -> Vec<(Place<'tcx>, Option<Self::Path>, Option<Local>)> {
+        if dataflow::move_path_children_matching(self.ctxt.move_data(), path,
+                                                 |_| true).is_none() {
+            return vec![];
+        }
+
+        assert!(size <= (u32::MAX as u64),
+                "move out check doesn't implemented for array bigger then u32");
+
+        let size = size as u32;
+        let flags = self.ctxt.array_items_drop_flags.get(&path);
+        (0..size).map(|i| {
+            let move_path = dataflow::move_path_children_matching(self.ctxt.move_data(), path, |p|
+                match p.elem {
+                    ProjectionElem::ConstantIndex{offset, min_length: _, from_end: false} =>
+                        offset == i,
+                    ProjectionElem::ConstantIndex{offset, min_length: _, from_end: true} =>
+                        size - offset == i,
+                    ProjectionElem::Subslice{from, to} => from <= i && i < size - to,
+                    _ => false
+                }
+            );
+            let place = &self.ctxt.move_data().move_paths[path].place;
+            (place.clone().elem(ProjectionElem::ConstantIndex{
+                offset: i,
+                min_length: size,
+                from_end: false
+            }),
+            move_path,
```

Comment:
So reading this, I'm afraid that this can cause a drop to be run with a "wrong" `MovePath` - a `Subslice` when the drop can also be accessed through a `ConstantIndex` (or vice versa), which might cause the "unconditional" drop code to trigger when it shouldn't, as in

```Rust
let mut x = [a.alloc(), a.alloc(), a.alloc()];
// The MP for x[1..3] is now initialized, x[1] is now uninitialized
let [_, _, _x] = x;
// this shouldn't rely on any drop flags, and just drop x[0] & x[1]
x = [a.alloc(), a.alloc(), a.alloc()];
let [_, .._x] = x;
// The MP for x[1..3] is now uninitialized, x[1] is now initialized
// end of scope drop for x - should drop x[0] without any drop flags
```

Reply:
And in some cases you might need the drop flags:
```Rust
let alloc = || (a.alloc(). a.alloc());
let mut x = [alloc(), alloc(), alloc()];
// The MP for x[1..3] is now initialized, x[2].0 is now uninitialized
let [_, _, (_x,_)] = x;
// this shouldn't rely on any drop flags, and just drop x[0], x[1], and x[2].1
x = [a.alloc(), a.alloc(), a.alloc()];
let [_, .._x] = x;
// The MP for x[1..3] is now uninitialized, x[1].1 is now initialized
// end of scope drop for x - should drop x[0] without any drop flags
```

So I think you *do* need to generate move paths for all the array elements when you see a subslice, to have a place to focus (similarly, you want to normalize the move paths so that an access to `array[-2 of 3]` is the same as an access to `array[1 of 3]`)

Reply:
At least, this is how it appears to me. There might be a better solution but it elides me.

Reply:
I'll go back to european time and no meetings in Monday (I'm currently in the US for the Mozilla All Hands), so if you need any help I'll go on IRC then.